Updated phpstan to 0.12.33, ninjify/qa to 0.11

adaamz opened the pull request.

mabar commented on .phpstan/composer.json:
```diff
@@ -0,0 +1,11 @@
+{
+    "require-dev": {
+        "phpstan/phpstan": "0.11.3",
+        "phpstan/phpstan-nette": "0.11.0"
```
`^0.11`

mabar commented on ruleset.xml:
```diff
@@ -8,11 +8,12 @@
                 src=>Contributte\ImageStorage,
-				tests/cases=>Tests\Cases
+                tests/cases=>Tests\Cases,
+                .phpstan/src=>Contributte\ImageStorage\\PHPStan
```
Single backslash

mabar commented on src/Image.php:
```diff
@@ -25,13 +25,16 @@ class Image
 
-	public function __construct(bool $friendly_url, string $data_dir, string $data_path, string $identifier, ?Traversable $props = null)
+	/**
+	 * @param iterable<string, bool|string|ImageNameScript|null>|null $props
```
Please don't use generics syntax until PSR-5 generics are accepted. There is missing support in a lot of tools for that.

adaamz replied:
slevomat/cs should support this already, phpstorm supports this already

mabar replied:
I tested it few days ago with phpstorm 2018.3 and it was only available for writing generics, no typehint when using variables with generics.

adaamz replied:
no typehint? what do you mean by that? there is no support in php itself.

mabar commented on src/ImageStorage.php:
```diff
@@ -22,10 +22,10 @@ class ImageStorage
 	private $data_dir;
 
-	/** @var string */
+	/** @var callable(string): string */
```
Interesting syntax. Where is it from? I found it only in [psalm docs](https://github.com/vimeo/psalm/blob/master/docs/typing_in_psalm.md#callables-and-closures)

adaamz replied:
i think originally from typescript, phpstan supports this some time, but it is undocumented

adaamz replied:
it was implemented in 0.10.3
https://github.com/phpstan/phpstan/releases/tag/0.10.3

adaamz committed:
updated to phpstan ^0.12

f3l1x commented on .phpstan/src/Type/ImageStorageGetNoImageMethodReturnTypeExtension.php:
```diff
@@ -0,0 +1,61 @@
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+
+final class ImageStorageGetNoImageMethodReturnTypeExtension implements DynamicMethodReturnTypeExtension
```
Uh, thank you. But I think it's not needed. We don't create custom extensions for phpstan for single package. I mean, if we do, we have to do it for all packages to be consistence across all packages.

adaamz replied:
This is needed for level 5+, because this method returns union. Previously it had php typehint for 1 type, which should cause type error from php.

adaamz replied:
btw users of this library can include this extension in their phpstan build and benefit from it too - less false positive/no ignore error because of union

paveljanda replied:
Thanks a lot for the PR but I would also consider removing this extension as it would require more maintenance. Thx

adaamz replied:
as you wish

f3l1x commented:
Thanks @adaamz for PR, could you please simplify it? Also remove all todo(s) from code?

adaamz commented:
@f3l1x
- removed todo
- updated phpstan to latest 0.11(.19), and used features from that
- rebased onto master

f3l1x commented:
Hi. Please @adaamz, let's upgrade phpstan to 0.12, move phpstan to main composer.json, drop custom phpstan extension.

adaamz commented:
@f3l1x @mabar 
I
- updated phpstan to 0.12
- add info about phpstan extension to composer.json (section `extra`)
- updated ninjify/qa (with according cs changes)

paveljanda reviewed (changes requested):
Could you please remove the ImageStorageGetNoImageMethodReturnTypeExtension? Thanks a lot!

adaamz commented:
removed phpstan dynamic return type extension, rebased onto master

f3l1x commented:
Thank you @adaamz
